Fix dummy interface returning changed

haddystuff opened the pull request:
##### SUMMARY
connection_options work differently when detecting changes for 802-3-ethernet.mtu parameter. It causes problems with applying mtu setting.
Fixes #3612

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE

- Bugfix Pull Request


##### COMPONENT NAME
plugins/modules/net_tools/nmcli.py

##### ADDITIONAL INFORMATION
After little small discussion(#3615) i came out with this change. Basically it shouldn't break previous behavior, but also should fix #3612.
This how it should work: When we try to run playbook with type "dummy" and no mtu parameter setted ansible won't do any changes and also will display no diff, but after you set(any way possible) mtu to some value except 0, next time you gonna run playbook without mtu it will change it to "auto" and will also show diff.

haddystuff commented:
Any comments regarding my last commit?

Ajpantuso commented on plugins/modules/net_tools/nmcli.py:
```diff
@@ -1695,6 +1695,11 @@ def _compare_conn_params(self, conn_info, options):
                     current_value = current_value.strip('"')
+                if key == self.mtu_setting and self.mtu is None:
+                    self.mtu = 0
+            elif key == self.mtu_setting:
+                # if self.mtu_setting parameter doesn't exist, NetworkManager behave like it's set to 'auto'
+                current_value = 'auto'
```
```suggestion
```
I don't think this part is useful and can be removed. This just makes the diff show `"auto"` instead of `null` after supplying a value for `mtu` with dummy connections when it was not previously applied. Which is confusing since the value was in fact `null`.

Ajpantuso reviewed (changes requested):
One comment below. You will also need to add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) as well.

haddystuff commented:
This part fixes the original bug #3612. If you have never changed mtu on an interface, the next time you would run playbook you will see pointless diff. For example, if i delete suggested lines and run this playbook against some host, that doesn't have dummy connection configured:
```---
- name: check dummy interface
  hosts: [redhat]
  become: true
  tasks:
    - name: Adding new dummy connection without mtu
      community.general.nmcli:
        conn_name: dummy_con0
        ifname: dummy0
        ip4: 10.0.0.1/32
        state: present
        type: dummy
    - name: Applying same config 
      community.general.nmcli:
        conn_name: dummy_con0
        ifname: dummy0
        ip4: 10.0.0.1/32
        state: present
        type: dummy
    - name: Applying same config again to check if problem still appears
      community.general.nmcli:
        conn_name: dummy_con0
        ifname: dummy0
        ip4: 10.0.0.1/32
        state: present
        type: dummy
```

i will get this output:
```
[alex@carla ansible]$ ansible-playbook bad_playbook.yaml --diff

PLAY [set dummy interface] *******************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
ok: [redhat]

TASK [Adding new dummy connection without mtu] ***********************************************************************************************************************************************************************************************
changed: [redhat]

TASK [Applying same config] ******************************************************************************************************************************************************************************************************************
--- before
+++ after
@@ -1,5 +1,5 @@
 {
-    "802-3-ethernet.mtu": null,
+    "802-3-ethernet.mtu": "auto",
     "connection.autoconnect": "yes",
     "connection.interface-name": "dummy0",
     "ipv4.addresses": "10.0.0.1/32",

changed: [redhat]

TASK [Applying same config again to check if problem still appears] **************************************************************************************************************************************************************************
--- before
+++ after
@@ -1,5 +1,5 @@
 {
-    "802-3-ethernet.mtu": null,
+    "802-3-ethernet.mtu": "auto",
     "connection.autoconnect": "yes",
     "connection.interface-name": "dummy0",
     "ipv4.addresses": "10.0.0.1/32",

changed: [redhat]

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
redhat                     : ok=4    changed=3    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
We can also replace ```current_value = 'auto'``` with ```continue```. But seem like it's more informative to have ```auto``` value in some outputs instead of just skipping it.

Ajpantuso commented:
> We can also replace `current_value = 'auto'` with `continue`. But seem like it's more informative to have `auto` value in some outputs instead of just skipping it.

Ok, to address the idempotency issues and preserve `null` in the diff I propose the following:

Remove the `elif` to set `current_value = None` and instead add the following after Line 1706:
```python
            elif all([key == self.mtu_setting, self.type == 'dummy', current_value is None, value == 'auto']):
                value = None
```

This produces the following for me:
```
TASK [Create dummy connection] **********************************************************************************************************************************************
changed: [localhost] => {"Connection": "Connection dummy_con0 of Type dummy is being added", "changed": true, "conn_name": "dummy_con0", "state": "present", "stdout": "Connection 'dummy_con0' (ac7a4a47-b156-4412-a956-d7e97983ea6b) successfully added.\n", "stdout_lines": ["Connection 'dummy_con0' (ac7a4a47-b156-4412-a956-d7e97983ea6b) successfully added."]}

TASK [Retrieve MTU value from nmcli] ****************************************************************************************************************************************
changed: [localhost] => {"changed": true, "cmd": ["nmcli", "-f", "802-3-ethernet.mtu", "conn", "show", "dummy_con0"], "delta": "0:00:00.020061", "end": "2021-11-01 12:26:21.636531", "msg": "", "rc": 0, "start": "2021-11-01 12:26:21.616470", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [Output MTU value] *****************************************************************************************************************************************************
ok: [localhost] => {
    "msg": ""
}

TASK [Create dummy connection] **********************************************************************************************************************************************
ok: [localhost] => {"Exists": "Connections already exist and no changes made", "changed": false, "conn_name": "dummy_con0", "state": "present"}

TASK [Retrieve MTU value from nmcli] ****************************************************************************************************************************************
changed: [localhost] => {"changed": true, "cmd": ["nmcli", "-f", "802-3-ethernet.mtu", "conn", "show", "dummy_con0"], "delta": "0:00:00.019151", "end": "2021-11-01 12:26:22.132594", "msg": "", "rc": 0, "start": "2021-11-01 12:26:22.113443", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [Output MTU value] *****************************************************************************************************************************************************
ok: [localhost] => {
    "msg": ""
}

TASK [Modify dummy connection with MTU of 1500] *****************************************************************************************************************************
--- before
+++ after
@@ -1,5 +1,5 @@
 {
-    "802-3-ethernet.mtu": null,
+    "802-3-ethernet.mtu": "1500",
     "connection.autoconnect": "yes",
     "connection.interface-name": "dummy0",
     "ipv4.addresses": "10.0.0.1/32",

changed: [localhost] => {"Exists": "Connections do exist so we are modifying them", "changed": true, "conn_name": "dummy_con0", "state": "present"}

TASK [Retrieve MTU value from nmcli] ****************************************************************************************************************************************
changed: [localhost] => {"changed": true, "cmd": ["nmcli", "-f", "802-3-ethernet.mtu", "conn", "show", "dummy_con0"], "delta": "0:00:00.019054", "end": "2021-11-01 12:26:22.676576", "msg": "", "rc": 0, "start": "2021-11-01 12:26:22.657522", "stderr": "", "stderr_lines": [], "stdout": "802-3-ethernet.mtu:                     1500", "stdout_lines": ["802-3-ethernet.mtu:                     1500"]}

TASK [Output MTU value] *****************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "802-3-ethernet.mtu:                     1500"
}

TASK [Modify dummy connection omitting MTU] *********************************************************************************************************************************
--- before
+++ after
@@ -1,5 +1,5 @@
 {
-    "802-3-ethernet.mtu": "1500",
+    "802-3-ethernet.mtu": "auto",
     "connection.autoconnect": "yes",
     "connection.interface-name": "dummy0",
     "ipv4.addresses": "10.0.0.1/32",

changed: [localhost] => {"Exists": "Connections do exist so we are modifying them", "changed": true, "conn_name": "dummy_con0", "state": "present"}

TASK [Retrieve MTU value from nmcli] ****************************************************************************************************************************************
changed: [localhost] => {"changed": true, "cmd": ["nmcli", "-f", "802-3-ethernet.mtu", "conn", "show", "dummy_con0"], "delta": "0:00:00.018994", "end": "2021-11-01 12:26:23.188803", "msg": "", "rc": 0, "start": "2021-11-01 12:26:23.169809", "stderr": "", "stderr_lines": [], "stdout": "802-3-ethernet.mtu:                     auto", "stdout_lines": ["802-3-ethernet.mtu:                     auto"]}

TASK [Output MTU value] *****************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "802-3-ethernet.mtu:                     auto"
}
```

haddystuff commented:
There's one problem with that change:

if i run this playbook:
```
---
- name: check dummy interface
  hosts: [redhat]
  become: true
  tasks:
    - name: Adding new dummy connection without mtu
      community.general.nmcli:
        conn_name: dummy_con0
        ifname: dummy0
        ip4: 10.0.0.1/32
        state: present
        type: dummy
    
    - name: Want to set mtu to auto 
      community.general.nmcli:
        conn_name: dummy_con0
        ifname: dummy0
        ip4: 10.0.0.1/32
        state: present
        type: dummy
        mtu: 0
    
    - name: Retrieve MTU value from nmcli
      ansible.builtin.command:
        cmd: nmcli -f '802-3-ethernet.mtu' conn show dummy_con0
      register: mtu_value
    - name: Output MTU value
      debug:
        msg: "{{ mtu_value.stdout }}"
    
    - name: Applying same config
      community.general.nmcli:
        conn_name: dummy_con0
        ifname: dummy0
        ip4: 10.0.0.1/32
        state: present
        type: dummy
```
i get this:
```
[alex@carla ansible]$ ansible-playbook bad_playbook.yaml --diff

PLAY [check dummy interface] *****************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
ok: [redhat]

TASK [Adding new dummy connection without mtu] ***********************************************************************************************************************************************************************************************
ok: [redhat]

TASK [Applying same config] ******************************************************************************************************************************************************************************************************************
ok: [redhat]

TASK [Retrieve MTU value from nmcli] *********************************************************************************************************************************************************************************************************
changed: [redhat]

TASK [Output MTU value] **********************************************************************************************************************************************************************************************************************
ok: [redhat] => {
    "msg": ""
}

TASK [Applying same config again to check if problem still appears] **************************************************************************************************************************************************************************
ok: [redhat]

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
redhat                     : ok=6    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```
As you can see, i just wanted to set  mtu to ```auto``` instead of null but i failed.
And if you run task with ```mtu: 0``` against some connection with mtu configured to let's say 1500, you will set mtu to auto as expected. So idempotency is broken here.

Maybe we can overcome previously mentioned problem, but final solution will look quite bulky(compare to ```current_value = 'auto'``` solution), and also will be little bit incorrect when showing that ```value``` is null, when actual ```value``` is auto. Solution with  ```current_value = 'auto'``` has similar problem, it shows that ```current_value``` is auto when it's null.

All in all, I think both solutions are lying a little bit when showing diff, but the one with  ```current_value = 'auto'``` looks potentially less buggy.

Ajpantuso commented:
> All in all, I think both solutions are lying a little bit when showing diff, but the one with `current_value = 'auto'` looks potentially less buggy.

My problem with manipulating `current_value` in this way is that it makes an assumption about `nmcli` that may change in the future and may lead to additional bifurcations in the code to support version based behavior.

Manipulating the `Ansible` value is ugly because of module's current state, but is more controllable on our end.

Adding `self.mtu is None` to the conditions will prevent the issue you identified when moving from `Null` to `auto`:
```python
            elif all([key == self.mtu_setting, self.type == 'dummy', current_value is None, value == 'auto', self.mtu is None]):
                value = None
```

haddystuff commented:
Also, i think we don't need to use ```self.type == 'dummy'``` in any elif statement.
First of all: other types works well with this changes.
Secondly: what if nmcli devs will come up with idea of adding new type of interface that behaves the same way in terms of mtu? We will need to add ```self.type == 'some_new_type'``` to elif statement.

Ajpantuso commented:
> Also, i think we don't need to use `self.type == 'dummy'` in any elif statement. First of all: other types works well with this changes. Secondly: what if nmcli devs will come up with idea of adding new type of interface that behaves the same way in terms of mtu? We will need to add `self.type == 'some_new_type'` to elif statement.

At the moment we know `dummy` is an edge case so I think it's better to be specific because there is a third scenario where a new interface behaves differently to either of the known scenarios and this creates more confusion. If a new interface behaves just like `dummy` then we reference the previous issue and refactor this condition into a method instead of having a complex list of `and`s.

haddystuff commented:
I mostly agree, but have some questions:
1. Do we still need ```value == 'auto'``` check?
```value``` is set to ```auto``` in ```connection_options```. But it can be important in terms ````connection_options``` behavior will be changed.

2. If i understand it right you propose change here:
```
if isinstance(current_value, list) and isinstance(value, list):
    # compare values between two lists
    if sorted(current_value) != sorted(value):
        changed = True
elif all([key == self.mtu_setting, self.type == 'dummy', current_value is None, value == 'auto', self.mtu is None]):
    value = None
else:
    if current_value != to_text(value):
        changed = True
```
Purpose of piece of code is to check if change has occurred. ```elif``` we want to add looks kinda strange here.

It seem to me that it would look less confusing if we send new line to it's own rule right before ```if isinstance(current_value, list) and isinstance(value, list):``` line, or insert ```elif``` to previous if statement and also add  ```current_value = None``` to execution.

Ajpantuso commented:
1. Yes, if you don't check for `value == auto` then this would take effect for any `mtu` value so you would not see a change when going from `Null` to a non-zero `mtu` if it was removed. (If you want to ensure the name is consistent I would make it a class constant instead)
2. The new conditional has two purposes. First to bypass change evaluation and second to set `value = None` to ensure the `diff` is accurate. You could add a line before to just handle `value = None`, but then the second `if else` block would have to be generalized to handle `current_value is None and value is None` which should be safe for the moment since empty values are skipped for any other setting. I don't like modifying the first block because that is really meant to transform `nmcli`'s output (we have to do it to set `self.mtu = 0`, but that's really a hack to avoid a breaking change)

haddystuff commented:
Seems right, should i write some additional tests for this bug to be checked #3612

Ajpantuso commented:
> Seems right. Should i write some additional tests for this bug to be checked #3612?

Additional tests would be greatly appreciated.

Ajpantuso reviewed (changes requested):
This just needs a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) and then it's good to go!

felixfontein commented on changelogs/fragments/3625-nmcli_false_changed_mtu_fix.yml:
```diff
@@ -0,0 +1,4 @@
+---
+bugfixes:
+  - nmcli - fixed falsely reported changed status when ``mtu`` is omitted with ``dummy`` connections
+    (https://github.com/ansible-collections/community.general/issues/3612, https://github.com/ansible-collections/community.general/pull/3625)
```
```suggestion
    (https://github.com/ansible-collections/community.general/issues/3612, https://github.com/ansible-collections/community.general/pull/3625).
```

Ajpantuso reviewed (approved):
LGTM

felixfontein commented:
@haddystuff thanks for fixing this!
@Ajpantuso thanks for reviewing!
